Update the reference

ehuss opened the pull request:
This updates the reference to use the new mdbook-spec preprocessor, which is a Cargo library inside the reference submodule.

Note that this PR contains a bunch of bootstrap cleanup commits to assist with making sure the submodules are working correctly. All of the cleanup PRs should have a description in their commit. I'd be happy to move those to a separate PR if that makes review easier.

The main changes for the reference are:
- Move the `doc::Reference` bootstrap step out of the generic macro into a custom step.
    - This step needs to build rustdoc because the new mdbook-spec plugin uses rustdoc for generating links.
    - PATH is updated so that the rustdoc binary can be found.
- rustbook now includes the mdbook-spec plugin as a dependency.
- rustbook enables the mdbook-spec preprocessor.

I did a bunch of testing with the various commands and setups, such as:
- `submodules=true` and `submodules=false`
- having all submodules deinitialized
- not in a git repository

However, there are probably thousands of different permutations of different commands, settings, and environments, so there is a chance I'm missing something.

ehuss committed:
Remove comment about bootstrap.py handling submodules

bootstrap.py handling of submodules was removed in
https://github.com/rust-lang/rust/pull/97513.

ehuss committed:
Remove `pub` from update_existing_submodules

This is not used anywhere outside this module.

ehuss committed:
Clarify comment on update_existing_submodules

This felt like an important point to me.

ehuss committed:
Remove argument from the submodules method

The argument was not necessary, since it was only ever passed one
value that exists in the config itself.

ehuss committed:
Remove outdated comment about update_submodule

Although its origins were in bootstrap.py, that code in bootstrap.py
no longer exists since it was removed.

ehuss committed:
Fix rustbook submodule update location

I put this submodule update in the entirely wrong location. I put it in
the `RustcBook` step (for generating src/doc/rustc), when it really
should exist for all steps that use the `Rustbook` tool.

ehuss committed:
Add require_and_update_submodule to ensure submodules exist

This adds a new method `require_and_update_submodule` to replace
`update_submodule`. This new method will generate an error if the
submodule doesn't actually exist. This replaces some ad-hoc checks that
were performing this function. This helps ensure that a good error
message is always displayed.

This also adds require_and_update_all_submodules which does this for
all submodules.

Ideally this should not have any change other than better error messages
when submodules are missing.

ehuss committed:
Clarify comment about why bootstrap tests need src/doc/book

ehuss committed:
Make sure submodules are checked out with `x test`

If the submodule is not checked out, then these tests would fail.

ehuss committed:
Integrate mdbook-spec for the reference.

This updates the reference which is now using a new mdbook plugin. This
requires a little extra work than a normal book because the plugin uses
`rustdoc` to generate links to the standard library. It also ensures
that the submodule is available for *any* command that uses rustbook,
since it is now part of the rustbook workspace.

JoelMarcey reviewed (approved):
LGTM - thanks!

Kobzol reviewed (commented):
Thanks! I went through the commits and left a few comments.

I tried to run `x doc reference` locally, and I got this:
```
Finished `release` profile [optimized] target(s) in 31.00s
Rustbook (x86_64-unknown-linux-gnu) - reference
error: manifest path `mdbook-spec/Cargo.toml` does not exist
```
when trying to run `x doc reference` for the first time, not sure what to make of it. Additional executions ended successfully (or at least it seemed like so).

One thing I also noticed is that the reference does not open automatically in the browser if I do `x doc reference` or `x doc src/doc/reference`, I have to use `--open` (not sure if this is correct or not, from the code it looked like it *might* be intended to be opened by default). But this also happens on `master`, so nothing to do with this PR.

ehuss committed:
Fix mistake setting ONLY_HOSTS for Reference.

This was a copy/paste mistake.

ehuss committed:
Add more descriptions to why submodules are required.

ehuss committed:
Consistently use a string to represent a submodule.

This makes it easier to call these functions without needing to form a
Path.

ehuss committed:
Change the blanket submodule update for library submodules to be required

These are required 100% of the time, but they are almost always required
for any command that runs Cargo in the main workspace.

Ideally, initializing these two standard library submodules would be
lazy and only initialized when required (see
https://github.com/rust-lang/rust/pull/82653). However, it would require
updating these in almost every Step (anything that runs `cargo` in the
main workspace).

ehuss committed:
Change prebuilt_llvm_config to not be required.

I misread this one. It is only checking if LLVM needs to be rebuilt.
There is code below that handles the case where it is unable to compute
the stamp if the source is missing.

ehuss committed:
Add clarifying documentation to require_and_update_submodule.

ehuss committed:
Rename require_and_update_submodule to require_submodule

Just trying to be a little less verbose here.

ehuss commented on src/bootstrap/src/core/build_steps/compile.rs:
```diff
@@ -456,13 +459,15 @@ pub fn std_cargo(builder: &Builder<'_>, target: TargetSelection, stage: u32, car
+        builder.require_and_update_submodule(
+            "src/llvm-project",
+            Some(
+                "need LLVM sources available to build `compiler-rt`, but they weren't present; \
```
I normalized the error message text here.

ehuss commented on src/bootstrap/src/core/build_steps/llvm.rs:
```diff
@@ -1270,7 +1270,7 @@ impl Step for Libunwind {
     /// Build libunwind.a
     fn run(self, builder: &Builder<'_>) -> Self::Output {
-        builder.update_submodule(Path::new("src/llvm-project"));
+        builder.require_and_update_submodule("src/llvm-project", None);
```
Sure, I added some suggestions!

ehuss commented on src/bootstrap/src/lib.rs:
```diff
@@ -473,7 +473,13 @@ impl Build {
+    /// tarball). Typically [`Build::require_and_update_submodule`] should be
+    /// used instead to provide a nice error to the user if the submodule is
+    /// missing.
+    fn update_submodule(&self, relative_path: &Path) {
```
I was on the fence for that, since it is not 100% required for *all* commands. However, it is required for the vast majority of commands, so I went ahead and changed it to make it required as you suggest.

Ideally, initializing these two standard library submodules would be lazy and only initialized when required (see https://github.com/rust-lang/rust/pull/82653). However, it would require updating these in almost every Step (anything that runs `cargo` in the main workspace).

ehuss commented on src/bootstrap/src/lib.rs:
```diff
@@ -579,6 +585,45 @@ impl Build {
+    ///
+    /// The given `err_hint` will be shown to the user if the submodule is not
+    /// checked out.
+    pub fn require_and_update_submodule(&self, submodule: &str, err_hint: Option<&str>) {
```
Sure!

ehuss commented on src/bootstrap/src/lib.rs:
```diff
@@ -579,6 +585,45 @@ impl Build {
+    ///
+    /// The given `err_hint` will be shown to the user if the submodule is not
+    /// checked out.
+    pub fn require_and_update_submodule(&self, submodule: &str, err_hint: Option<&str>) {
```
Sure! Yea, I was on the fence of naming it that way. My thinking was to reinforce the idea that it does two different things. But shorter is good, too.

ehuss commented on src/bootstrap/src/lib.rs:
```diff
@@ -579,6 +585,45 @@ impl Build {
+            } else {
+                ""
+            };
+            let err_hint = err_hint.map_or_else(String::new, |e| format!("\n{e}"));
```
Sorry, just to clarify here. There is already a newline being prepended (so that it is on a line by itself). Are you saying there should be a second newline so there is a blank line separating them?

ehuss replied:
For context, the error currently looks like this:

```
submodule library/backtrace does not appear to be checked out, but it is required for this step
Consider setting `build.submodules = true` or manually initializing the submodules.
The submodule is required for the standard library and the main Cargo workspace.
Build completed unsuccessfully in 0:00:00
```

Kobzol replied:
Sorry, I somehow read `map_or_else` as `unwrap_or_else` :man_facepalming: Nevermind.

ehuss commented:
> when trying to run `x doc reference` for the first time, not sure what to make of it. Additional executions ended successfully (or at least it seemed like so).

That's an unfortunate consequence of how the mdbook preprocessors are overridden. The error is ignored, and it builds successfully. I don't have a simple way to silence it right now. That's what [this comment](https://github.com/rust-lang/rust/blob/53ef052d455ba8bd9a4823c59696efb257aef2fa/src/tools/rustbook/src/main.rs#L87-L93) is referring to about being unable to suppress the error message. I will try to get mdbook updated in a future PR soon so that it replaces the existing preprocessor to avoid that.

> One thing I also noticed is that the reference does not open automatically in the browser if I do `x doc reference` or `x doc src/doc/reference`, I have to use `--open` (not sure if this is correct or not, from the code it looked like it _might_ be intended to be opened by default). But this also happens on `master`, so nothing to do with this PR.

That's normal. It is required to use `--open` to open it automatically.

----

BTW, while re-reviewing, I noticed a mistake in the `prebuilt_llvm_config` function. It should not *require* the submodule. There is code below when computing the stamp to handle the case when the submodule source is missing.

Kobzol commented:
> That's normal. It is required to use --open to open it automatically.

Sorry, I did not notice [this condition](https://github.com/rust-lang/rust/blob/6062059ab0f9a805ae8327a80e55bc6f47330a7b/src/bootstrap/src/core/builder.rs#L2342).

Thank you!

@bors r+ rollup=iffy

bors commented:
:pushpin: Commit 1c98b8f9226d0e7a4e0ef42dd558de2857812f7a has been approved by `Kobzol`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Kobzol 1c98b8f9226d0e7a4e0ef42dd558de2857812f7a -->
<!-- homu: {"type":"Approved","sha":"1c98b8f9226d0e7a4e0ef42dd558de2857812f7a","approver":"Kobzol","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 1c98b8f9226d0e7a4e0ef42dd558de2857812f7a with merge 78c857394ec8c01f06cb1df260c51178180a40e5...
<!-- homu: {"type":"BuildStarted","head_sha":"1c98b8f9226d0e7a4e0ef42dd558de2857812f7a","merge_sha":"78c857394ec8c01f06cb1df260c51178180a40e5"} -->
